### PR TITLE
Added typing and missing CP7 functions

### PR DESCRIPTION
**Pull request by makermelissa**
This updates the API to the latest CircuitPython 7 and adds Typing to everything.

### CODE REVIEW

**Comment by FoamyGuy** [comment]
@makermelissa what do you think about updating the main Blinka project to allow it to try to continue even on unsupported environments. I.e. print warning instead of raise exception here inside of microcontroller/__init__.py:

https://github.com/FoamyGuy/Adafruit_Blinka/blob/d5f2a94f96e5eec35ea16bb08dc059005051407d/src/microcontroller/__init__.py#L141

I created a branch with that change and can open a PR with it if you're open to this idea.

I attempted to test the changes from this PR using https://github.com/FoamyGuy/Blinka_Displayio_PyGameDisplay on Linux PC but noticed that the version from this branch raises this error:
```
Traceback (most recent call last):
  File "/home/timc/repos/circuitpython/Blinka_Displayio_PyGameDisplay/examples/blinka_displayio_pygamedisplay_simpletest.py", line 8, in <module>
    import displayio
  File "/home/timc/.local/lib/python3.8/site-packages/displayio/__init__.py", line 26, in <module>
    from ._display import Display
  File "/home/timc/.local/lib/python3.8/site-packages/displayio/_display.py", line 27, in <module>
    import microcontroller
  File "/home/timc/.local/lib/python3.8/site-packages/microcontroller/__init__.py", line 141, in <module>
    raise NotImplementedError("Microcontroller not supported:", chip_id)
NotImplementedError: ('Microcontroller not supported:', 'GENERIC_X86')
```
Because of the newly added `import microcontroller` statements which are used for typing info.

Or another potential solution is to add another branch to the if statement to allow `GENERIC_X86` and simply pass without importing anything further. Like this:
```py
elif chip_id == ap_chip.GENERIC_X86:
    print("WARNING: GENERIC_X86 is not fully supported. Some features may not work.")
    pass
```

I tested both of these possible solutions with the Blinka_Displayio_PyGameDisplay library on my Linux PC and both do all it to run successfully. 

I think it's worth printing the warning in either case because the vast majority of functionality that Blinka provides is not intended to work in that environment but `displayio` specifically can because of the PyGameDisplay library. I have done all of my testing on Linux, but ideally it would be great if we can allow it to import succesfully on Windows and Mac also which might require adding more branches to the if statement if we go with the 2nd solution above. But I think with the first solution (changing raise to print) it would work as-is on windows and mac no matter what they end up getting as their `chip_id`

File: displayio/_epaperdisplay.py [hunk 1]
```diff
@@ -36,38 +41,38 @@ class EPaperDisplay:
...
+        set_row_window_command: Optional[int] = None,
+        set_current_column_command: Optional[int] = None,
+        set_current_row_command: Optional[int] = None,
+        write_black_ram_command: int,
+        black_bits_inverted: bool = False,
+        write_color_ram_command: Optional[int] = None,
+        color_bits_inverted: bool = False,
+        highlight_color: int = 0x000000,
+        refresh_display_command: int,
+        refresh_time: float = 40,
```
Comment (FoamyGuy) [inline displayio/_epaperdisplay.py:65]: Should be `int` or possible the default should be `40.0` ?

File: displayio/_epaperdisplay.py [hunk 2]
```diff
@@ -36,38 +41,38 @@ class EPaperDisplay:
...
+        busy_pin: Optional[microcontroller.Pin] = None,
+        busy_state: bool = True,
+        seconds_per_frame: float = 180,
```
Comment (FoamyGuy) [inline displayio/_epaperdisplay.py:68]: Same for this one should it be `int` or possibly `180.0` ?

File: displayio/_group.py [hunk 3]
```diff
@@ -108,38 +105,38 @@ def insert(self, index, layer):
...
         index = self.index(layer)
         self._layers.pop(index)
 
-    def __len__(self):
+    def __len__(self) -> int:
         """Returns the number of layers in a Group"""
         return len(self._layers)
 
-    def __getitem__(self, index):
+    def __getitem__(self, index) -> Union[Group, TileGrid]:
```
Comment (FoamyGuy) [inline displayio/_group.py:129]: could have `index: int` not sure if this is necessary for the builtin __getitem__ function though? maybe it's okay to leave it out.

File: displayio/_group.py [hunk 4]
```diff
@@ -108,38 +105,38 @@ def insert(self, index, layer):
...
         """Returns the value at the given index."""
         return self._layers[index]
 
-    def __setitem__(self, index, value):
+    def __setitem__(self, index, value) -> None:
```
Comment (FoamyGuy) [inline displayio/_group.py:133]: could have `index: int` and `value: Union[Group, TileGrid]`

File: displayio/_group.py [hunk 5]
```diff
@@ -108,38 +105,38 @@ def insert(self, index, layer):
...
         """Sets the value at the given index."""
         self._removal_cleanup(index)
         self._layers[index] = value
         self._layer_update(index)
 
-    def __delitem__(self, index):
+    def __delitem__(self, index) -> None:
```
Comment (FoamyGuy) [inline displayio/_group.py:139]: could have `index: int`

**Review by FoamyGuy — COMMENTED** [review]
Thanks for working on this update to match current APIs as well as refactoring and adding the type information @makermelissa 

I found a few spots that could possibly have more type hints added. 

I looked over all of the changes in this PR and most everything is looking good to me. I would like to try to get a solution for the issue mentioned about GENERIC_X86 in place before merging if possible so that the PyGameDisplay library can continue working.

File: displayio/_epaperdisplay.py (hunk 1 above)
Comment (makermelissa) [reply displayio/_epaperdisplay.py:65]: I copied off https://circuitpython.readthedocs.io/en/latest/shared-bindings/displayio/index.html#displayio.EPaperDisplay, but I'm fine with adding a .0 to the end.

File: displayio/_group.py (hunk 3 above)
Comment (makermelissa) [reply displayio/_group.py:129]: Yep, looks like I missed that one.

File: displayio/_group.py (hunk 4 above)
Comment (makermelissa) [reply displayio/_group.py:133]: Missed that one too

**Comment by makermelissa** [comment]
@FoamyGuy I'm fine with you adding the GENERIC_X86 branch as a PR if you want. I'll get the other changes made.

**Review by FoamyGuy — APPROVED** [review]
I made [#533](https://github.com/adafruit/Adafruit_Blinka/pull/533) for the `GENERIC_X86` allowance.

The changes here look good to. Thanks again!

**Pull request merged**